### PR TITLE
[Merged by Bors] - chore(Tactic/Polyrith): remove Python `requests` dependency in SageMath call

### PR DESCRIPTION
**Pull request by hanwenzhu**
*   Replaced the `requests` library needed by the Python script for SageMath API call, with native library `urllib.request`.
    *   E.g. if the user has no `pip install` rights, such as evaluation on a shared server
*   Added explicit `User-Agent` header to SageMath requests: `LeanProver (https://leanprover-community.github.io/)`. Previously, it was the default used by Python requests package `python-requests/...`.

---

I am not completely sure this works in all cases where requests worked.

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by hanwenzhu** [comment]
I would also argue perhaps removing dependency on Python entirely makes sense, and we can move the logic inside Lean except for the HTTP call itself (maybe use something like `Cache.IO.runCurl`?)

File: scripts/polyrith_sage.py
Comment (grunweg) [inline scripts/polyrith_sage.py:71]: Clarifying to my future self: `urlencode` returns an ASCII string, and that must be encoded to bytes before using it in a request. Actually, reading the documentation --- shouldn't you encode this to bytes, using e.g. `bytes(url, 'utf-8')` where (`url` is `urllib.parse.urlencode({'code': query})`)?

> For an HTTP POST request method, data should be a buffer in the standard application/x-www-form-urlencoded format. The [urllib.parse.urlencode()](https://docs.python.org/3/library/urllib.parse.html#urllib.parse.urlencode) function takes a mapping or sequence of 2-tuples and returns an ASCII string in this format. It should be encoded to bytes before being used as the data parameter.

File: scripts/polyrith_sage.py
Comment (grunweg) [inline scripts/polyrith_sage.py:72]: Optional: capitalise the headers as `Content-Type` and `User-Agent`? (This is only a stylistic question.)

File: Mathlib/Tactic/Polyrith.lean
Comment (grunweg) [inline Mathlib/Tactic/Polyrith.lean:393]: Actually, `request` apparently required Python 3.8 --- so I would definitely increase the requirement here.
Given that Python 3.8 is end-of-live, I'd be happy increasing the requirement to 3.9.

That said: you're not checking that version anywhere, are you?

**Review by grunweg — APPROVED** [review]
The general shape looks reasonable to me, and this seems like a nice clean-up. Thanks for the PR!
I left a few detail questions inline.

**Comment by grunweg** [comment]
In principle, rewriting Python into Lean always seems nice. I am not sure to what extend the HTTP operations are supported; that's something you'd have to find out. (There are some threads on zulip about this. Advanced features are not supported, but this PR only needs basic things.)

File: scripts/polyrith_sage.py
Comment (hanwenzhu) [reply scripts/polyrith_sage.py:71]: That's correct, I did encode the string into bytes using `.encode('utf-8')`. That is equivalent to `bytes(..., 'utf-8')` as for as I know.

File: Mathlib/Tactic/Polyrith.lean
Comment (hanwenzhu) [reply Mathlib/Tactic/Polyrith.lean:393]: I am not checking that version anywhere.

I don't think request requires Python 3.8. `urllib.request` was added with Python 3.

Comment (hanwenzhu) [reply Mathlib/Tactic/Polyrith.lean:393]: If you mean the package `requests` that was removed, it has [versions](https://pypi.org/project/requests/2.17.3/) for Python all the way back to Python 2. Although probably those versions are not actively maintained anymore.

**Comment by hanwenzhu** [comment]
> In principle, rewriting Python into Lean always seems nice. I am not sure to what extend the HTTP operations are supported; that's something you'd have to find out. (There are some threads on zulip about this. Advanced features are not supported, but this PR only needs basic things.)

The simplest way to do this is just via a call to `curl`, like in Cache or LeanSearchClient. I can look into that later, but probably in a separate PR

File: Mathlib/Tactic/Polyrith.lean
Comment (grunweg) [reply Mathlib/Tactic/Polyrith.lean:393]: In other words: you could also say "has Python 3 installed" --- which is deliberately unclear, but under most readings does not include end-of-life'd versions. Perhaps that's better; what do you think?

Comment (hanwenzhu) [reply Mathlib/Tactic/Polyrith.lean:393]: Ah, definitely that makes sense!

Comment (hanwenzhu) [reply Mathlib/Tactic/Polyrith.lean]: The documentation is now reverted to `python3` installed and available on path

**Review by grunweg — COMMENTED** [review]
Assuming the `bytes` comment is correct (didn't double-check), this looks good to me. Thanks!
maintainer merge

**Comment by hanwenzhu** [comment]
I am also working on a PR to completely remove dependency on Python. It seems like a pretty small change

**Comment by hanwenzhu** [comment]
Thanks! @grunweg

**Comment by kmill** [comment]
Thanks!

It looks correct to me, but it's been ages since I've used `requests`, so I'd like to wait a little bit before merging in case anyone knows if there's a corner case that this `urllib`-ification is missing. I'll delegate merging to you, but please don't do `bors r+` until the following conditions are met:
1. Either 12 hours have passed or another maintainer gives this PR their 'ok'; and
2. You do one last test locally that polyrith still works with this change. I assume you tested it, but it would be good to have it on the record that it was tested.

bors d+

**Comment by kmill** [comment]
I just had a moment to test it. Seems to work!

bors r+

**Pull request closed without merging**
